inmoose/pycombat/covariates.py

```diff
@@ -23,274 +23,372 @@
+            self.batch_mod = dmatrix("~0 + C(self.batch)")
+
+        # reference batch
+        if ref_batch is not None:
+            print(type(ref_batch))
+            if ref_batch not in self.batch.categories:
+                raise ValueError("Reference batch must identify one of the batches")
+            # ref_batch_idx is the index of the reference batch in batch.categories
+            self.ref_batch_idx = np.where(self.batch.categories == ref_batch)[0][0]
+            # update batchmod with reference
+            self.batch_mod[:, self.ref_batch_idx] = 1
+        else:
+            self.ref_batch_idx = None
+
+        if covar_mod is not None and not isinstance(covar_mod, DesignMatrix):
+            # convert to dataframe
+            if isinstance(covar_mod, (list, np.ndarray)):
+                covar_mod = pd.DataFrame(covar_mod)
+                if covar_mod.shape[0] != len(self.batch):
+                    if covar_mod.shape[1] == len(self.batch):
+                        covar_mod = covar_mod.T
+                    else:
+                        raise ValueError(
+                            "The covar_mod matrix parameter doesn't match the number of sample"
+                        )
+                covar_mod.columns = [
+                    "cov_" + str(col_nb) for col_nb in covar_mod.columns
+                ]
+            elif type(covar_mod) != pd.DataFrame:
+                raise ValueError(
+                    f"The covar_mod parameter type {type(covar_mod)} is not accepted, it must be list, numpy.array or pandas.dataframe."
+                )
 
-class ConfoundingVariablesError(Exception):
-    """Exception raised when confounding variables are detected.
+            nan_covar_mod = covar_mod.isna()
+            self.nan_cov = nan_covar_mod.any(axis=0).sum()
+            self.nan_cov_samples = {
+                b: nan_covar_mod[self.batch == b].any(axis=0).sum()
+                for b in self.batch.categories
+            }
+        else:
+            self.nan_cov = 0
+            self.nan_cov_samples = {b: 0 for b in self.batch.categories}
+
+        if self.nan_cov == 0:
+            if covar_mod is not None:
+                if not isinstance(covar_mod, DesignMatrix):
+                    covar_mod = dmatrix(
+                        "+".join([f"{cv}" for cv in covar_mod.columns]),
+                        data=covar_mod,
+                    )
+            else:
+                covar_mod = dmatrix("~1", pd.DataFrame(counts.T))
+
+            self.confounded_cov = []
+            for c in range(covar_mod.shape[1]):
+                # skip the intercept
+                if (covar_mod[:, c] == 1).all():
+                    continue
+                if (
+                    np.linalg.matrix_rank(
+                        np.hstack([self.batch_mod, covar_mod[:, c][:, None]])
+                    )
+                    <= self.batch_mod.shape[1]
+                ):
+                    self.confounded_cov.append(c)
```

Comment:
Not sure it is possible, but what about variables confounded because of linear combination of 3 or more columns?